File: understading/user_data.py
"""
here I want to understand
how each user see their data basend on their acount
"""

# data
note = [
  {
    "id": 1,
    "title": "My first note",
    "content": "This is the content of my first note.",
    "user_id": 1
  },
  {
    "id": 2,
    "title": "My second note",
    "content": "This is the content of my second note.",
    "user_id": 1
  },
  {
    "id": 3,
    "title": "My third note",
    "content": "This is the content of my third note.",
    "user_id": 2
  }
]

def display_note_i(note_i):
  """
  Function to display a single note.
  """
  print(f"Note ID: {note_i['id']}")
  print(f"Title: {note_i['title']}")
  print(f"Content: {note_i['content']}")
  print(f"User ID: {note_i['user_id']}")
  print("-" * 20)

def display_notes(notes):
  """
  Function to display notes.
  """
  for note in notes:
    display_note_i(note)

def view_note_by_user(user_id):
  """
  Function to view notes based on user_id.
  """
  user_notes = [
    note_i 
    for note_i in note 
    if note_i["user_id"] == user_id
  ]
  return user_notes

data = view_note_by_user(2)
display_notes(data)

File: understading/user_data.c
#include <stdio.h>
#include <string.h>

/*=========================================
    Structure
=========================================*/

typedef struct
{
  const char *id;
  const char *user_id;
  const char *content;

} Note;

/*=========================================
    Database (Temporary)
=========================================*/

const Note notes[] =
{
  {"n001", "u001", "This is the first note."},
  {"n002", "u002", "This is the second note."},
  {"n003", "u001", "This is another note for user u001."},
  {"n004", "u003", "This is a note for user u003."},
  {"n005", "u002", "This is another note for user u002."},
  {"n006", "u001", "This is yet another note for user u001."},
  {"n007", "u003", "This is a second note for user u003."},
  {"n008", "u002", "This is a third note for user u002."},
  {"n009", "u001", "This is a fourth note for user u001."},
  {"n010", "u003", "This is a third note for user u003."}
};

const int note_count = sizeof(notes) / sizeof(notes[0]);

/*=========================================
    Display One Note
=========================================*/

void display_note(const Note *note)
{
  printf("Note ID : %s\n", note->id);
  printf("User ID : %s\n", note->user_id);
  printf("Content : %s\n", note->content);
  printf("---------------------------------\n");
}

/*=========================================
    Display Notes By User
=========================================*/

int display_notes_by_user(
  const Note *notes,
  int note_count,
  const char *target_user_id)
{
  int found = 0;

  for (int i = 0; i < note_count; i++)
  {
    if (strcmp(notes[i].user_id, target_user_id) == 0)
    {
      display_note(&notes[i]);
      found++;
    }
  }

  return found;
}

/*=========================================
    Main
=========================================*/

int main(void)
{
  const char *target_user_id = "u003";

  printf("=================================\n");
  printf("Private Notes\n");
  printf("=================================\n\n");

  printf("Current User : %s\n\n", target_user_id);

  int total = display_notes_by_user(
    notes,
    note_count,
    target_user_id
  );

  if (total == 0)
  {
    printf("No notes found.\n");
  }
  else
  {
    printf("Total Notes : %d\n", total);
  }
  
  return 0;
}